apiserver/tools.go

```diff
@@ -352,60 +495,63 @@ func (h *toolsUploadHandler) getServerRoot(r *http.Request, query url.Values, st
 func (h *toolsUploadHandler) handleUpload(
 	ctx context.Context,
 	r io.Reader,
-	toolsVersions []semversion.Binary,
-	serverRoot string,
-	st *state.State,
-	store objectstore.ObjectStore,
-) (*tools.Tools, error) {
-	serviceFactory, err := h.ctxt.domainServicesForRequest(ctx)
+	agentBinaryStore AgentBinaryStore,
+	agentBinaryVersion coreagentbinary.Version,
+) (int64, string, error) {
+	blockChecker, err := h.blockCheckerGetter(ctx)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return 0, "", internalerrors.Errorf(
+			"failed getting block checker for tools upload request: %w", err,
+		)
 	}
 
-	// Check if changes are allowed and the command may proceed.
-	blockChecker := common.NewBlockChecker(serviceFactory.BlockCommand())
 	if err := blockChecker.ChangeAllowed(ctx); err != nil {
-		return nil, errors.Trace(err)
+		return 0, "", internalerrors.Capture(err)
 	}
-	storage, err := st.ToolsStorage(store)
-	if err != nil {
-		return nil, err
-	}
-	defer storage.Close()
 
 	// Read the tools tarball from the request, calculating the sha256 along the way.
 	data, sha256, size, err := tmpCacheAndHash(r)
 	if err != nil {
-		return nil, err
+		return 0, "", internalerrors.Errorf("caching and hashing agent binary upload: %w", err)
 	}
 	defer data.Close()
 
 	if size == 0 {
-		return nil, errors.BadRequestf("no agent binaries uploaded")
-	}
-
-	// TODO(wallyworld): check integrity of tools tarball.
-
-	// Store tools and metadata in tools storage.
-	for _, v := range toolsVersions {
-		metadata := binarystorage.Metadata{
-			Version: v.String(),
-			Size:    size,
-			SHA256:  sha256,
-		}
-		logger.Debugf(context.TODO(), "uploading agent binaries %+v to storage", metadata)
-		if err := storage.Add(ctx, data, metadata); err != nil {
-			return nil, err
-		}
-	}
-
-	tools := &tools.Tools{
-		Version: toolsVersions[0],
-		Size:    size,
-		SHA256:  sha256,
-		URL:     common.ToolsURL(serverRoot, toolsVersions[0].String()),
-	}
-	return tools, nil
+		return 0, "", internalerrors.New("no agent binaries uploaded").Add(coreerrors.BadRequest)
+	}
+
+	// TODO(wallyworld, tlm): check integrity of tools tarball. This todo was
+	// added before the integration of Dqlite into this handler. What we ideally
+	// should be doing is letting the agent binary store disect the upload if we
+	// wish for this to be done.
+
+	logger.Debugf(
+		ctx,
+		"uploading agent binaries for version %q and arch %q to agent binary store",
```

Comment:
Can we include the sha as well?

Reply:
We can.

Reply:
Done & Done

Reply:
Is there a commit to push? Don't see the sha?